crates/sui-indexer-alt/src/pipeline/concurrent/collector.rs

```diff
@@ -179,3 +182,166 @@ pub(super) fn collector<H: Handler + 'static>(
+
+        let batch2 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch2.len(), 2);
+
+        let batch3 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch3.len(), 0);
+
+        cancel.cancel();
+    }
+
+    #[tokio::test]
+    async fn test_collector_shutdown() {
+        let (processor_tx, processor_rx) = mpsc::channel(10);
+        let (collector_tx, mut collector_rx) = mpsc::channel(10);
+        let metrics = Arc::new(IndexerMetrics::new(&Registry::new()));
+        let cancel = CancellationToken::new();
+
+        let _collector = collector::<TestHandler>(
+            PipelineConfig::default(),
+            processor_rx,
+            collector_tx,
+            metrics,
+            cancel.clone(),
+        );
+
+        processor_tx
+            .send(Indexed::new(0, 1, 10, 1000, vec![1, 2]))
+            .await
+            .unwrap();
+
+        let batch = collector_rx.recv().await.unwrap();
+        assert_eq!(batch.len(), 2);
+
+        // Drop processor sender to simulate shutdown
+        drop(processor_tx);
+
+        // After a short delay, collector should shut down
+        tokio::time::sleep(Duration::from_millis(200)).await;
+        assert!(collector_rx.try_recv().is_err());
+
+        cancel.cancel();
+    }
+
+    #[tokio::test]
+    async fn test_collector_respects_max_pending() {
+        let processor_channel_size = 5; // unit is checkpoint
+        let collector_channel_size = 10; // unit is batch, aka rows / MAX_CHUNK_ROWS
+        let (processor_tx, processor_rx) = mpsc::channel(processor_channel_size);
+        let (collector_tx, _collector_rx) = mpsc::channel(collector_channel_size);
+
+        let metrics = Arc::new(IndexerMetrics::new(&Registry::new()));
+
+        let cancel = CancellationToken::new();
+
+        let _collector = collector::<TestHandler>(
+            PipelineConfig::default(),
+            processor_rx,
+            collector_tx,
+            metrics.clone(),
+            cancel.clone(),
+        );
+
+        // Send more data than MAX_PENDING_ROWS plus collector channel buffer
+        let data = Indexed::new(
+            0,
+            1,
+            10,
+            1000,
+            vec![
+                1;
+                TestHandler::MAX_PENDING_ROWS
+                    + TestHandler::MAX_CHUNK_ROWS * collector_channel_size
+            ],
+        );
+        processor_tx.send(data).await.unwrap();
+
+        // Now fill up the processor channel with minimum data to trigger send blocking
+        for _ in 0..processor_channel_size {
+            let more_data = Indexed::new(0, 2, 11, 1000, vec![1]);
+            processor_tx.send(more_data).await.unwrap();
+        }
+
+        // Now sending even more data should block.
+        let even_more_data = Indexed::new(0, 3, 12, 1000, vec![1]);
+
+        let send_result = tokio::time::timeout(
+            Duration::from_millis(2000),
+            processor_tx.send(even_more_data),
```

Comment:
I wonder if you can run this test by checking the length of the channel instead? That way you won't have to wait 2s each time.

Reply:
Changed this to using `try_send` and asserting on error type.